HopsanGUI/Widgets/ModelWidget.cpp

```diff
@@ -698,16 +698,254 @@ void ModelWidget::saveAs()
+        ssvUnit = dataElement.attribute(ssv::attr::unit);
+        ssvValue = dataElement.attribute(ssv::attr::value);
+
+        //Now find corresponding parameter in model
+        //ssvName = system1|system2|component.parameter (component parameters)
+        //ssvName = system1|system2|parameter           (system paramters)
+        QStringList systemHierarchy;
+        QString componentName, parameterName;
+        systemHierarchy = ssvName.split("|");
+        if(systemHierarchy.last().contains(".")) {
+            componentName = systemHierarchy.last().split(".")[0];  //Component parameter
+            parameterName = systemHierarchy.last().split(".")[1];
+        }
+        else {
+            parameterName = systemHierarchy.last();                 //System parameter
+        }
+        systemHierarchy.removeLast();
+        SystemObject *pSystem = getTopLevelSystemContainer();
+        for(int s=0; s<systemHierarchy.size(); ++s) {               //Recurse subsystems
+            pSystem = qobject_cast<SystemObject*>(pSystem->getModelObject(systemHierarchy[s]));
+            if(pSystem == nullptr) {
+                gpMessageHandler->addWarningMessage("Could not find subsystem: "+systemHierarchy[s]+", ignoring all its parameters.");
+                parameterElement = parameterElement.nextSiblingElement(ssv::parameter);
+                continue;
+            }
+        }
+        if(componentName.isEmpty()) {                               //Set system parameter
+            CoreParameterData parameter;
+            pSystem->getParameter(parameterName, parameter);
+            if(ssv::dataTypeTranslator[parameter.mType] != ssvDataType) {
+                gpMessageHandler->addWarningMessage("Wrong data type for unit: \""+ssvName+"\", ignoring value. ("+ssv::dataTypeTranslator[parameter.mType]+" != "+ssvDataType+")");
+            }
+            else if(parameter.mType == ssv::datatype::real) {    //Handle units for "Real" type parameters only
+                if(parameter.mUnit == ssvUnit) {
+                    pSystem->setParameterValue(parameterName, ssvValue);
+                }
+                else {
+                    QStringList quantities = gpConfig->getQuantitiesForUnit(ssvUnit);
+                    if(quantities.isEmpty()) {
+                        gpMessageHandler->addWarningMessage("Unit is not supported by Hopsan: "+ssvUnit+". Setting parameter without unit conversion.");
+                        pSystem->setParameterValue(parameterName, ssvValue);
+                    }
+                    else {
+                        QString ssvQuantity, hopsanUnit, hopsanQuantity;
+                        ssvQuantity = quantities[0];    //Guessing for first quantity is the best we can do
+                        hopsanUnit = parameter.mUnit;
+                        if(parameter.mQuantity.isEmpty()) {
+                            hopsanQuantity = gpConfig->getQuantitiesForUnit(parameter.mUnit)[0];
```

Comment:
Must make sure returned list is not empty before taking [0], could be that someone creates a component and make up some custom unit

Reply:
Same as above, we never get here if quantities is empty.

Reply:
but in this case we retrieve quantities for parameter.mUnit, as far as I can see above, we get quantities for ssvUnit
They are not the same in this case due to the if check at line 873. (This check did not exist in the other place)
Also I noticed now that hopsanUnit is assigned but not used, (as far as I can see in the diff)

Reply:
Added a check here too.